### PR TITLE
Kernel: Properly implement ControlMemory FREE and COMMIT

### PR DESCRIPTION
**Pull request by yuriks**
This gets rid of another hard-coded part of the address space.

### CODE REVIEW

File: src/core/hle/kernel/process.cpp [hunk 1]
```diff
@@ -28,6 +28,109 @@ SharedPtr<CodeSet> CodeSet::Create(std::string name, u64 program_id) {
...
+
+    CASCADE_RESULT(auto vma, vm_manager.MapMemoryBlock(target, heap_memory, target - heap_start, size, MemoryState::Private));
+    vm_manager.Reprotect(vma, perms);
+
+    return MakeResult<VAddr>(heap_end - size);
+}
+
+ResultCode ProcessMemoryManager::HeapFree(VAddr target, u32 size) {
+    if (target < Memory::HEAP_VADDR || target + size > Memory::HEAP_VADDR_END || target + size < target) {
+        return ERR_INVALID_ADDRESS;
```
Comment (archshift) [inline src/core/hle/kernel/process.cpp]: Maybe add a log if this happens so we can know (if the game, for whatever reason, doesn't error-check this)?

File: src/core/hle/kernel/process.cpp [hunk 2]
```diff
@@ -28,6 +28,109 @@ SharedPtr<CodeSet> CodeSet::Create(std::string name, u64 program_id) {
...
+
+    VAddr heap_end = Memory::LINEAR_HEAP_VADDR + (u32)linear_heap_memory->size();
+    // Games and homebrew only ever seem to pass 0 here (which lets the kernel decide the address),
+    // but explicit addresses are also accepted and respected.
+    if (target == 0) target = heap_end;
+
+    if (target < Memory::LINEAR_HEAP_VADDR || target + size > Memory::LINEAR_HEAP_VADDR_END ||
+        target > heap_end || target + size < target) {
+
+        return ERR_INVALID_ADDRESS;
```
Comment (archshift) [inline src/core/hle/kernel/process.cpp]: Ditto.

File: src/core/hle/kernel/process.cpp [hunk 3]
```diff
@@ -28,6 +28,109 @@ SharedPtr<CodeSet> CodeSet::Create(std::string name, u64 program_id) {
 CodeSet::CodeSet() {}
 CodeSet::~CodeSet() {}
 
+ResultVal<VAddr> ProcessMemoryManager::HeapAllocate(VAddr target, u32 size, VMAPermission perms) {
+    if (target < Memory::HEAP_VADDR || target + size > Memory::HEAP_VADDR_END || target + size < target) {
+        return ERR_INVALID_ADDRESS;
```
Comment (archshift) [inline src/core/hle/kernel/process.cpp]: Ditto.

**Comment by Subv** [comment]
I'm not sure if it's related to these changes, but any game that calls GetSharedFont will crash citra due to GetPointer returning nullptr

**Comment by yuriks** [comment]
It sure is. `SHARED_FONT_VADDR` is 0x18000000, which is somewhere in the middle of the linear heap area, which now does not have a dummy mapping anymore. I was expecting the font to be mapped in the shared memory area at 0x10000000-0x14000000. We need to find out why it's positioned where it currently is.

**Comment by yuriks** [comment]
I need to do some testing around this and the patch will need some impovements to properly support whatever's going on here, so I'm going to close it for now.

**Pull request closed without merging**